File: Remoter-Mac/Sources/RemoterAgent/Config/AgentConfig.swift
import Foundation

// Persisted config — ~/Library/Application Support/Remoter/config.json
// CLI args always take precedence over this file.
struct AgentConfig: Codable {
    var port: UInt16 = 7788
    var pin: String  = ""
    var relayUrl: String = ""
    // TURN（coturn）共享密钥，首次启动随机生成（见 TurnCredentials.ensureSecret），
    // 不写死在源码里——这个文件不进 git，跟 pin 一样只留在本机。
    var turnSecret: String = ""

    private static let fileURL: URL = {
        let dir = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Remoter", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("config.json")
    }()

    static func load() -> AgentConfig {
        guard let data = try? Data(contentsOf: fileURL),
              let cfg  = try? JSONDecoder().decode(AgentConfig.self, from: data)
        else { return AgentConfig() }
        return cfg
    }

    func save() {
        guard let data = try? JSONEncoder().encode(self) else { return }
        try? data.write(to: AgentConfig.fileURL, options: .atomic)
    }
}

File: Remoter-Win/Config/AgentConfig.cs
using System.Text.Json;

namespace RemoterWin;

// Persisted config — %LOCALAPPDATA%\Remoter\config.json
// CLI args always take precedence over this file.
sealed class AgentConfig
{
    public ushort Port     { get; set; } = 7788;
    public string Pin      { get; set; } = "";
    public string RelayUrl { get; set; } = "";

    private static readonly string _path = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "Remoter", "config.json");
    private static readonly object _lock = new();

    public static AgentConfig Load()
    {
        lock (_lock)
        {
            try
            {
                if (!File.Exists(_path)) return new();
                var json = File.ReadAllText(_path);
                return JsonSerializer.Deserialize<AgentConfig>(json,
                    new JsonSerializerOptions { PropertyNameCaseInsensitive = true }) ?? new();
            }
            catch { return new(); }
        }
    }

    public void Save()
    {
        lock (_lock)
        {
            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
                File.WriteAllText(_path, JsonSerializer.Serialize(this,
                    new JsonSerializerOptions { WriteIndented = true }));
            }
            catch { }
        }
    }
}
